### PR TITLE
Allow usage of .tar.gz with PretrainedModelInitializer

### PR DESCRIPTION
**Pull request by eladsegal**

### CODE REVIEW

**Comment by dirkgr** [comment]
This is a nice feature, but there is quite a bit of overlap with what `cached_path()`'s "`extract_archive`" feature already does. Is there a way we can just use `cached_path()`? It will extract the archive into the cache directory, which means it needs no cleanup, and therefore no context manager.

**Comment by eladsegal** [comment]
> This is a nice feature, but there is quite a bit of overlap with what `cached_path()`'s "`extract_archive`" feature already does. Is there a way we can just use `cached_path()`? It will extract the archive into the cache directory, which means it needs no cleanup, and therefore no context manager.

Personally I'm ok with no cleanup, but see these recent issue and the corresponding pull request - https://github.com/allenai/allennlp/issues/4642, https://github.com/allenai/allennlp/pull/4643.

File: tests/nn/pretrained_model_initializer_test.py [hunk 1]
```diff
@@ -79,6 +81,23 @@ def test_from_params(self):
...
+    def test_from_params_tar_gz(self):
+        try:
+            # Create temp.tar.gz
+            temp_tar_gz = os.path.join(self.TEST_DIR, "temp.tar.gz")
+            with tarfile.open(temp_tar_gz, "w:gz") as archive:
+                archive.add(self.temp_file, arcname=os.path.basename(self.temp_file))
+            params = Params({"type": "pretrained", "weights_file_path": temp_tar_gz})
+            initializer = Initializer.from_params(params)
+        finally:
+            os.remove(temp_tar_gz)
```
Comment (dirkgr) [inline tests/nn/pretrained_model_initializer_test.py:93]: Why not `with tempfile.NamedTemporaryFile(...):`?

This is an excellent place for a context manager :-)

**Review by dirkgr — CHANGES_REQUESTED** [review]
The big issue is that `load_archive_contextmanager()` doesn't make sense in its current form. At least, I don't see it. Why do we need it?

The other comments are details.

**Comment by eladsegal** [comment]
> The big issue is that `load_archive_contextmanager()` doesn't make sense in its current form. At least, I don't see it. Why do we need it?
> 
> The other comments are details.

Hi @dirkgr, indeed using contextmanager doesn't make sense for this pull request.
I worked on multiple changes that were split across multiple pull request, so this change was made with another pull request in mind - https://github.com/allenai/allennlp/pull/4713.
When extracting an archive for the `evaluate` or `predict` commands, passing the extraction directory as serialization_dir to the model/dataset reader constructor must be done before the extraction directory is deleted - this is the justification I saw for contextmanager.
It is also why I return `archive, serialization_dir` from `load_archive_contextmanager`.
I could have returned only `archive` and add `extraction_path/serialization_dir` as a member in it, but it made less sense to me to have a member that holds an already deleted path.

Example of the usage:
https://github.com/allenai/allennlp/blob/8a8451d6c1120b543cc7ec1b2860c3dc154131c9/allennlp/commands/evaluate.py#L120-L139

File: allennlp/models/archival.py [hunk 2]
```diff
@@ -162,36 +192,58 @@ def load_archive(
...
-            weights_file=weights_path,
-            serialization_dir=serialization_dir,
-            cuda_device=cuda_device,
-        )
+            with extract_archive(resolved_archive_file, cleanup=False) as tempdir:
+                serialization_dir = tempdir
+                model, config = _load_model(cuda_device, overrides, weights_file, serialization_dir)
 
+        archive = Archive(model=model, config=config)
+        yield archive, serialization_dir
```
[the remark this answers is not in the document]
Comment (eladsegal) [reply allennlp/models/archival.py:202]: I agree, I don't like it too.
The issue is that archive is indeed valid afterwards, but serialization_dir is not.
I'll try to think how to split it in a cleaner way

**Comment by dirkgr** [comment]
Can you solve that scenario with something like this?
```Python
with extracted_archive(...) as d:
   archive = load_archive(d)
  do_other_stuff_with(d)
```

In your example, you do this because you want to pass the serialization directory to the reader. Can we make the reader part of `Archive` instead so that problem goes away?

I have a question open on the other PR about why the reader needs the archive in the first place.

**Comment by eladsegal** [comment]
> Can you solve that scenario with something like this?
> 
> ```python
> with extracted_archive(...) as d:
>    archive = load_archive(d)
>    do_other_stuff_with(d)
> ```
That's exactly what I did in the the other pull request :)

> 
> In your example, you do this because you want to pass the serialization directory to the reader. Can we make the reader part of `Archive` instead so that problem goes away?
>
You are right, it will solve the problem, and I think it is also reasonable to include the dataset reader object when loading an archive.
I'll make the change, Archive object will have two new fields: `dataset_reader` and `validation_dataset_reader` (so `Predictor` will be able to select `dataset_reader_to_load`).

**Comment by eladsegal** [comment]
@dirkgr I've made the requested changes

**Review by dirkgr — APPROVED** [review]
I have some comments, but it also looks good as it is. Let me know if you want to change something, otherwise I'll hit the merge button. Thank you for this one (and the other ones, including the ones still pending).

File: tests/models/archival_test.py [hunk 3]
```diff
@@ -43,13 +44,14 @@ def setup_method(self):
                 "train_data_path": str(self.FIXTURES_ROOT / "data" / "sequence_tagging.tsv"),
                 "validation_data_path": str(self.FIXTURES_ROOT / "data" / "sequence_tagging.tsv"),
                 "data_loader": {"batch_size": 2},
-                "trainer": {"num_epochs": 2, "optimizer": "adam"},
+                "trainer": {"num_epochs": 2, "optimizer": "adam", "cuda_device": -1},
```
[the remark this answers is not in the document]
Comment (eladsegal) [reply tests/models/archival_test.py:47]: When running on a machine without a GPU, the test will pass. However, on a machine with GPU, the trainer will automatically set `cuda_device` to 0 ([here](https://github.com/allenai/allennlp/blob/3a85e3595b71c05fbd9acb014e3d76b5b7dd0bc4/allennlp/training/trainer.py#L56-L62)) and the returned model will be on the GPU while the model from the archive will be on the CPU.

**Comment by eladsegal** [comment]
> I have some comments, but it also looks good as it is. Let me know if you want to change something, otherwise I'll hit the merge button. Thank you for this one (and the other ones, including the ones still pending)!

I've made the validation dataset reader, ready for merge.
And you are welcome, I'm happy to contribute!

**Comment by eladsegal** [comment]
Hi @dirkgr, can we merge this?

**Comment by dirkgr** [comment]
@eladsegal, yes, we can. I'm going on vacation beginning tomorrow. For the next two weeks, can you ping @epwalsh if there are things I forgot to take care of?

**Pull request merged**